### ISSUE TITLE
Npgsql 3.0.x does not return Postgis Geometry

### ISSUE THREAD

**Issue by keyjote** [nhibernate/NHibernate.Spatial#31]
I have been struggling for a few weeks when mapping my Geometry column in Postgis to IGeometry
here is a very detailed description of the [problem](http://gis.stackexchange.com/questions/178117/nhibernate-spatial-geometry-type-become-null)

The solution in the end was to downgrade Npsql to use 2.2.7 and it just works.

I tried to report this to Nhibernate jira bug tracker but seems I am not allowed to, I might need to be logged in to their jira system in order to do so. 
that is why I am reporting it here, maybe you guys can help me get this issue fixed.

**Comment by andrerav** [comment]
Hi @keyjote. For the sake of experimentation, can you try using Npgsql 3.1 alpha? It's available on NuGet. The Npgsql claim to have implemented support for PostGIS 2D types in that version. See here: https://github.com/npgsql/npgsql/issues/807

**Comment by keyjote** [comment]
Hi @andrerav. I updatd to Npgsql version 3.1.0-alpha6 and I got null for my geometry (Polygon)
If I look at the logs I see this:

`2016-02-19 08:01:33,883 [18] DEBUG NHibernate.Type.StringClobType [(null)] - returning 'NpgsqlTypes.PostgisPolygon' as column: the10_22_0_`

which I would assume it did work (previously, this line would say: returning null as column_ the10_22_0_) but when trying to use the property Geom in the code I get it to be null.

Anything in particular I should check?

**Comment by andrerav** [comment]
@keyjote: After looking at the [code](https://github.com/nhibernate/NHibernate.Spatial/blob/master/NHibernate.Spatial.PostGis/Type/PostGisGeometryType.cs) it definitely seems that there's not much I can do in NHibernate.Spatial.PostGis about this. The problem stems from either NHibernate (more likely) or NetTopologySuite.IO.PostGis (less likely). Is it possible for you to share your code so I can get a test case up and running?

**Comment by keyjote** [comment]
@andrerav I don't think I will have the time right now to create a test project for you. I can describe my setup for now and over the weekend I can create a test project and then upload it here.

My setup uses NHibernate version 4.0.4.4000 and NHibernate.Spatial.PostGis 4.0.4.4000 (with Npgsql version 3.0.5, also tested with 3.1.0-alpha6) and running a Postgres db version 9.4 and Postgis version 2.1.7

in the web.config
`<property name="dialect">NHibernate.Spatial.Dialect.PostGisDialect,NHibernate.Spatial.PostGis</property>`
Note. I used this [fix](https://github.com/nhibernate/NHibernate.Spatial/issues/23) so for the Dialect so it supports IdentityColumns

for the nhibernate configuration I have
```
    var cfg = new NHibernate.Cfg.Configuration();
    cfg.AddAuxiliaryDatabaseObject(new SpatialAuxiliaryDatabaseObject(cfg));
```

my xml mapping file has the following mapping:
```
        <property name="Geom" column="the_geom">
            <type name="NHibernate.Spatial.Type.PostGisGeometryType, NHibernate.Spatial.PostGis">
                <param name="subtype">POLYGON</param>
                <param name="srid">3006</param>
            </type>
        </property> 

```
and my model class has the following property
`public virtual IGeometry Geom { get; set; }`

Hope this at least gives you some guidance, otherwise I will upload a demo project over the weekend.

**Comment by andrerav** [comment]
Thanks @keyjote, I'm leaving town for the weekend, but this info is enough for me to create a test case. I will take a look at this next week.

**Comment by keyjote** [comment]
@andrerav I appreciate it :)

**Comment by megafetis** [comment]
Problem with postgis and NHibernate is solved by creating correct mapping type
`[Serializable]
public class GeometryType : AbstractStringType
{
internal GeometryType() : base(new StringSqlType())
{
}

    public override object Get(IDataReader rs, int index)
    {
        int length = (int)rs.GetBytes(index, 0, null, 0, 0);
        byte[] buffer = new byte[length];
        if (length > 0)
        {
            rs.GetBytes(index, 0, buffer, 0, length);
        }
        return buffer;
    }

    public override object Get(IDataReader rs, string name)
    {
        return Get(rs, rs.GetOrdinal(name));
    }

    internal GeometryType(StringSqlType sqlType) : base(sqlType)
    {
    }

    public override string Name
    {
        get { return "Geometry"; }
    }
}

[Serializable]
public class PostGisGeometryTypeCorrect : Zaprosotvet.Models.GeometryTypeBase<string>
{
    public PostGisGeometryTypeCorrect(): base(new GeometryType())
    {
        
    }

    protected override void SetDefaultSRID(GeoAPI.Geometries.IGeometry geometry)
    {
        geometry.SRID = 4326;
    }
    protected override string FromGeometry(object value)
    {
        IGeometry geometry = value as IGeometry;
        if (geometry == null)
        {
            return null;
        }
        if (!(geometry is IGeometryCollection) && geometry.IsEmpty)
        {
            geometry = GeometryCollection.Empty;
        }

        this.SetDefaultSRID(geometry);

        var ordinates = Ordinates.XY;
        var interiorPoint = geometry.InteriorPoint;
        if (!interiorPoint.IsEmpty && !double.IsNaN(interiorPoint.Z))
        {
            ordinates |= Ordinates.Z;
        }
        if (!interiorPoint.IsEmpty && !double.IsNaN(interiorPoint.M))
        {
            ordinates |= Ordinates.M;
        }

        var postGisWriter = new PostGisWriter
        {
            HandleOrdinates = ordinates
        };
        byte[] bytes = postGisWriter.Write(geometry);
        return ToString(bytes);
    }
    protected override IGeometry ToGeometry(object value)
    {
        var byteArray = value as byte[];
        var b = ToString(byteArray);
        string bytes = b;

        if (string.IsNullOrEmpty(bytes))
        {
            return null;
        }

        // Bounding boxes are not serialized as hexadecimal string (?)
        const string boxToken = "BOX(";
        if (bytes.StartsWith(boxToken))
        {
            // TODO: Optimize?
            bytes = bytes.Substring(boxToken.Length, bytes.Length - boxToken.Length - 1);
            string[] parts = bytes.Split(',');
            string[] min = parts[0].Split(' ');
            string[] max = parts[1].Split(' ');
            string wkt = string.Format(
                "POLYGON(({0} {1},{0} {3},{2} {3},{2} {1},{0} {1}))",
                min[0], min[1], max[0], max[1]);
            return new WKTReader().Read(wkt);
        }

        PostGisReader reader = new PostGisReader();
        IGeometry geometry = reader.Read(bytes.StartsWith(boxToken) ? ToByteArray(bytes) : byteArray);
        this.SetDefaultSRID(geometry);
        return geometry;
    }

    private static byte[] ToByteArray(string hex)
    {
        if (hex.Length % 2 == 1)
            throw new ArgumentException("Invalid input. It must have an even length.", "hex");

        byte[] data = new byte[hex.Length / 2];
        for (int i = 0; i < hex.Length; i += 2)
        {
            char c1 = hex[i];
            char c2 = hex[i + 1];

            int result = (c1 < 'A') ? (c1 - '0') : (10 + (c1 - 'A'));
            result = result << 4;
            result |= (c2 < 'A') ? (c2 - '0') : (10 + (c2 - 'A'));
            data[i / 2] = (byte)(result);
        }
        return data;
    }

    private static string ToString(byte[] bytes)
    {
        char[] data = new char[bytes.Length * 2];
        int idx = 0;
        for (int i = 0; i < bytes.Length; i++)
        {
            int n1 = bytes[i] >> 4;
            int n2 = bytes[i] & 0xF;
            data[idx++] = (char)((n1) < 10 ? '0' + n1 : n1 - 10 + 'A');
            data[idx++] = (char)((n2) < 10 ? '0' + n2 : n2 - 10 + 'A');
        }
        return new string(data);
    }
}

[Serializable]
public abstract class GeometryTypeBase<T> : NHibernate.Spatial.Type.IGeometryUserType
{

    protected abstract T FromGeometry(object value);

    protected abstract IGeometry ToGeometry(object value);

    private int srid = -1;
    private string subtype = "GEOMETRY";
    private int dimension = 2;

    private readonly NullableType nullableType;
    private readonly SqlType sqlType;

    protected GeometryTypeBase(NullableType nullableType)
    {
        this.nullableType = nullableType;
        this.sqlType = nullableType.SqlType;
    }

    #region IGeometryUserType Members

    public object DeepCopy(object value)
    {
        var geom = (IGeometry)value;
        var bytes = geom.AsBinary();
        byte[] newArr = new byte[bytes.Length];
        bytes.CopyTo(newArr,0);
        return this.ToGeometry(newArr);
    }

    public object Disassemble(object value)
    {
        return value;
    }

    public object Assemble(object cached, object owner)
    {
        return cached;
    }

    public int GetHashCode(object x)
    {
        return x.GetHashCode();
    }

    public bool IsMutable
    {
        get { return true; }
    }

    public object NullSafeGet(IDataReader rs, string[] names, object owner)
    {
        try
        {
            var result = this.nullableType.NullSafeGet(rs, names);
            return this.ToGeometry(result);
        }
        catch
        {
            return null;
        }
    }

    public void NullSafeSet(IDbCommand cmd, object value, int index)
    {
        var geom = (IGeometry) value;
        geom.SRID = this.SRID;
        var fromGeom = FromGeometry(geom);
        this.nullableType.NullSafeSet(cmd, fromGeom, index);
    }

    public object Replace(object original, object target, object owner)
    {
        return original;
    }

    public virtual System.Type ReturnedType
    {
        get { return typeof(Geometry); }
    }

    public virtual SqlType[] SqlTypes
    {
        get { return new SqlType[] { this.sqlType }; }
    }

    public override bool Equals(object obj)
    {
        return base.Equals(obj);
    }

    bool IUserType.Equals(object a, object b)
    {
        IGeometry ga = a as IGeometry;
        IGeometry gb = b as IGeometry;
        if (ga != null && gb != null)
        {
            try
            {
                return ga.SRID == gb.SRID && ga.EqualsExact(gb);
            }
            catch (TopologyException)
            {
                return false;
            }
        }
        if (ga == null && gb == null)
        {
            return NHibernate.Util.EqualsHelper.Equals((T)a, (T)b);
        }
        return false;
    }

    public override int GetHashCode()
    {
        return base.GetHashCode();
    }

    public void SetParameterValues(IDictionary<string, string> parameters)
    {
        if (parameters != null)
        {
            string parameterSRID = parameters["srid"];
            string parameterSubtype = parameters["subtype"];

            if (!string.IsNullOrEmpty(parameterSRID))
            {
                Int32.TryParse(parameterSRID, out this.srid);
            }
            if (!string.IsNullOrEmpty(parameterSubtype))
            {
                this.subtype = parameterSubtype;
            }
            //[3DIS]
            string strDimension;
            parameters.TryGetValue("dimension", out strDimension);
            if (!string.IsNullOrEmpty(strDimension))
            {
                Int32.TryParse(strDimension, out this.dimension);
            }
        }
    }

    public int SRID
    {
        get { return this.srid; }
    }

    public string Subtype
    {
        get { return this.subtype; }
    }

    public int Dimension
    {
        get { return this.dimension; }
    }

    /// <summary>
    /// Compares two objects and returns a value indicating whether one is less than, equal to, or greater than the other.
    /// </summary>
    /// <param name="x">The first object to compare.</param>
    /// <param name="y">The second object to compare.</param>
    /// <returns>
    /// Value Condition Less than zero <paramref name="x"/> is less than <paramref name="y"/>. Zero <paramref name="x"/> equals <paramref name="y"/>. Greater than zero <paramref name="x"/> is greater than <paramref name="y"/>.
    /// </returns>
    /// <exception cref="T:System.ArgumentException">Neither <paramref name="x"/> nor <paramref name="y"/> implements the <see cref="T:System.IComparable"/> interface.-or- <paramref name="x"/> and <paramref name="y"/> are of different types and neither one can handle comparisons with the other. </exception>
    public int Compare(object x, object y)
    {
        throw new InvalidOperationException("Invalid operation in geometry type");
    }

    #endregion IGeometryUserType Members

    protected virtual void SetDefaultSRID(IGeometry geometry)
    {
        if (geometry.SRID <= 0 && this.SRID > 0)
        {
            geometry.SRID = this.SRID;
        }
    }
}

Mapping IGeometry property:Property(p => p.Geometry, m =>
{
m.Type();
});`

property of the entity:
public virtual IGeometry Geometry { get; set; }

It works with Npgsql 3.2.5 and last versions of Nhibernate.Spatial

**Comment by peetw** [comment]
@megafetis Thanks for that! What are the main differences from the current code that make it work? (The formatting seems to be a bit off so I can't easily see, sorry!)

**Comment by megafetis** [comment]
I created a new type "GeometryType" based on AbstractStringType from Nhibernate lib. The "Get" method executes as BinaryType, returning binary geometry. 
The StringType returned "Nhpsqltype.PostgisGeometry" string. It was wrong. 
Then I created class GeometryTypeBase from IGeometryUserType interface as in source libruary, but with corrections in some methods. It initializes by passing GeometryType: 
```
public PostGisGeometryTypeCorrect(): base(new GeometryType())
{
// init
}

```
Then I created class `PostGisGeometryTypeCorrect : GeometryTypeBase`
It overrides ToGeometry method. There converts byte[] to IGeometry

I took source code from Nhibernate, Nhibernate.Spatial, Nhibernate.Spatial.Postgis

**Comment by megafetis** [comment]
[PostGisGeometryTypeCorrect.zip](https://github.com/nhibernate/NHibernate.Spatial/files/1204704/PostGisGeometryTypeCorrect.zip)
 my code

**Comment by DMDTT** [comment]
Baaaaahm,

Great I'm gonna check your code.
Maybe we finally can get rid of npgsql 2.2.7

**Comment by keyjote** [comment]
Great work @megafetis , I did some initial testing on your fix and seems to be working great using:
* npgsql: 3.2.5
* nhibernate: 4.1.1.4000 and 
* nhibernate.spatial.postgis: 4.1.1.4001

The only problem I had is that I do not use WGS84 for the coordinate system on my applications, so having geometry.SRID = 4326; in the PostGisGeometryTypeCorrect class broke my application when trying to save data.
To solve this issue I removed the method SetDefaultSRID and then I was able to save. I guess NHibernate gets the SRID from the geometry column and sets the value, so I don't see the need for this overload.

I will test a bit more this week and get back to you on my findings.
For now it is looking very promessing, thanks :)

**Comment by megafetis** [comment]
Ok, I am glad to help you :)

2017-08-29 10:04 GMT+03:00 Keyjote <notifications@github.com>:

**Comment by peetw** [comment]
@keyjote @megafetis thanks for looking into this guys! Would either/both of you be able to submit a PR with the necessary changes? Also, have you tested whether the changes work, even with the old 2.x version of npgsql?

**Comment by megafetis** [comment]
No, I doesn't yet

On 31 Aug 2017 11:36 am, "Peet Whittaker" <notifications@github.com> wrote:

@keyjote <https://github.com/keyjote> @megafetis
<https://github.com/megafetis> thanks for looking into this guys! Would
either/both of you be able to submit a PR with the necessary changes? Also,
have you tested whether the changes work, even with the old 2.x version of
npgsql?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/nhibernate/NHibernate.Spatial/issues/31#issuecomment-326229146>,
or mute the thread

**Comment by megafetis** [comment]
Excuse me for my bad English

**Comment by DMDTT** [comment]
Unfortunately this immediately makes the entity "dirty" for nhibernate, so if you load the entity and you don't persist the entity on transaction.commit an update will be fired for this entity

**Comment by megafetis** [comment]
What is beter solution?

2017-11-02 10:59 GMT+03:00 DMDTT <notifications@github.com>:

**Comment by DMDTT** [comment]
in fact I do not have one but adding a custom update listener


            bool somethingHasChanged = false;
            for (int index = 0; index < @event.State.Length; index++)
            {
                bool propertyHasChanged = (@event.OldState[index] == null && @event.State[index] != null)
                                          || (@event.OldState[index] != null && @event.State[index] == null)
                                          || (@event.OldState[index] != null && @event.State[index] != null && !@event.OldState[index].Equals(@event.State[index]));
                somethingHasChanged = somethingHasChanged || propertyHasChanged;
            }

            return !somethingHasChanged; //// negate this expression as on false an update of the entity will be executed

**Comment by maca88** [comment]
I took some time to analyze the current situation about the support for `Npgsql 3` and I figured out that there are actually three way of adding support for it:

1. The first one is similar to what @megafetis posted, which is to transform the binary data that can be retrived from the `DbDataReader` into a string. With an additinal condition we are also able to maintain the support for `Npgsql 2.x`. This approach works with `Npgsql 2.x` and with `Npgsql 3.2.5` and higher. [Here](https://github.com/maca88/NHibernate.Spatial/commit/882f4c60ae79ada8930b5699f6ae404eff556e04) are the changes for this approach. 

2. The second one is by using the binary data that can be retrived from the `DbDataReader` to represent the geometry types instead of strings. With this approach we have to drop the support for `Npgsql 2.x` and add a `Npgsql` reference to `NHibernate.Spatial.PostGis` project, in order to set the parameter type to `NpgsqlDbType.Geometry`. This approach works only with `Npgsql 3.2.5` and higher, as in version `3.2.5` they added support for [receiving geometry as byte array](https://github.com/npgsql/npgsql/pull/1643). [Here](https://github.com/maca88/NHibernate.Spatial/commit/71699e9e0690bb253475c834c654f2de625c54fa) are the changes for this approach. 

3. The last approach that I know is to replace the default `Npgsql` geometry type handler with a custom one, which will be supported in `Npgsql 3.3` (not yet released). This means that user will need to add an additional configuration line to replace the `Npgsql` handler and `Npgsql 3.3` would be the minimum supported version. Do note that with the current state of [this approach](https://github.com/maca88/NHibernate.Spatial/commit/14e0baca850a8d4d94deb98131b42535f0061e87) there are still failing tests that needs to be fixed.

To summarize the options:

First apporach:
*+* `Npgsql 2.x` support
*+* Minimal change
*-* The slowest
*-* Harder to maintain because of `2.x` support

Second approach:
*+* The easiest to maintain
*+* Minimal change
*-* `Npgsql` reference on `NHibernate.Spatial.PostGis` project
*-* No support for `Npgsql 2.x`

Third approach:
*+* The fastest
*-* Only for `Npgsql 3.3+`
*-* Additional configuration line
*-* `Npgsql` reference on `NHibernate.Spatial.PostGis` project
*-* Needs work as there are failing tests
*-* No support for `Npgsql 2.x`

Personally, I would choose between the first or second approach, probably the second one. What are your thoughts about this?

**Comment by peetw** [comment]
@maca88 thanks for looking into this and for the detailed post!

My initial reaction would probably be that the first approach is the most suitable as it requires the least changes and has the widest compatibility. I will have more time to give it a proper look through this weekend though.

**Comment by maca88** [comment]
> My initial reaction would probably be that the first approach is the most suitable

Would be fine for me, the only thing that is missing in the changes for the first approach, is the ability to running tests for both `Ngpsql 2.x` and `3.x`. Probably this will require to have two extra test projects.

**Comment by DMDTT** [comment]
I like the third one, but is the nhibernate spatial project still maintained?

I think the dropping npgsql 2.x support is worth a look, as it maintained for a long time

**Comment by peetw** [comment]
@maca88 Hmmm, yes, good point about the test projects. That certainly moves the argument slightly away from option 1 and towards options 2 and 3.

In fact, the more I think about it, the less I think that adding a reference to `Npgsql` from the `NHibernate.Spatial.PostGis` project is a problem - `Npgsql` is by far and away the most popular PostgreSQL driver for .NET and I can't really think of any other serious alternatives (speak up if you disagree!). So realistically, anyone using the `NHibernate.Spatial.PostGis` project is probably already using `Npgsql`. This leads me to think that option 2 may actually be the best solution. @andrerav @suryapratap thoughts?

@DMDTT The project is still maintained (ish), but I haven't had as much spare time to give to the project as I would wish recently. The main project admins are @suryapratap and @andrerav, but I think they are in a similar position to me at the moment (although I would love to be proved wrong :)).

As for your comment re: dropping support for `Npgsql` 2.x, I think that may be acceptable (see comments above).

**Comment by peetw** [comment]
I have now merged in the PR from @maca88 - hopefully we can get a new NuGet release done ASAP (dependent on @andrerav )
